compiler/rustc_mir_build/src/thir/pattern/usefulness.rs

```diff
@@ -1148,28 +1151,45 @@ pub(crate) fn compute_match_usefulness<'p, 'tcx>(
 
     // Run the non_exhaustive_omitted_patterns lint. Only run on refutable patterns to avoid hitting
     // `if let`s. Only run if the match is exhaustive otherwise the error is redundant.
-    if cx.refutable
-        && non_exhaustiveness_witnesses.is_empty()
-        && !matches!(
+    if cx.refutable && non_exhaustiveness_witnesses.is_empty() {
+        if !matches!(
             cx.tcx.lint_level_at_node(NON_EXHAUSTIVE_OMITTED_PATTERNS, lint_root).0,
             rustc_session::lint::Level::Allow
-        )
-    {
-        let witnesses = collect_nonexhaustive_missing_variants(cx, &pat_column);
-        if !witnesses.is_empty() {
-            // Report that a match of a `non_exhaustive` enum marked with `non_exhaustive_omitted_patterns`
-            // is not exhaustive enough.
-            //
-            // NB: The partner lint for structs lives in `compiler/rustc_hir_analysis/src/check/pat.rs`.
-            cx.tcx.emit_spanned_lint(
-                NON_EXHAUSTIVE_OMITTED_PATTERNS,
-                lint_root,
-                scrut_span,
-                NonExhaustiveOmittedPattern {
-                    scrut_ty,
-                    uncovered: Uncovered::new(scrut_span, cx, witnesses),
-                },
-            );
+        ) {
+            let witnesses = collect_nonexhaustive_missing_variants(cx, &pat_column);
+
+            if !witnesses.is_empty() {
+                // Report that a match of a `non_exhaustive` enum marked with `non_exhaustive_omitted_patterns`
+                // is not exhaustive enough.
+                //
+                // NB: The partner lint for structs lives in `compiler/rustc_hir_analysis/src/check/pat.rs`.
+                cx.tcx.emit_spanned_lint(
+                    NON_EXHAUSTIVE_OMITTED_PATTERNS,
+                    lint_root,
+                    scrut_span,
+                    NonExhaustiveOmittedPattern {
+                        scrut_ty,
+                        uncovered: Uncovered::new(scrut_span, cx, witnesses),
+                    },
+                );
+            }
+        } else {
+            // We used to allow putting the `#[allow(non_exhaustive_omitted_patterns)]` on a match
+            // arm. This no longer makes sense so we warn users, to avoid silently breaking their
+            // usage of the lint.
+            for arm in arms {
+                if !matches!(
+                    cx.tcx.lint_level_at_node(NON_EXHAUSTIVE_OMITTED_PATTERNS, arm.hir_id).0,
+                    rustc_session::lint::Level::Allow
+                ) {
+                    cx.tcx.emit_spanned_lint(
+                        NON_EXHAUSTIVE_OMITTED_PATTERNS,
+                        arm.hir_id,
+                        arm.pat.span(),
+                        NonExhaustiveOmittedPatternLintOnArm,
+                    );
```

Comment:
Should we add a suggestion to move the attribute on the match itself?

Reply:
Oh yes good idea, I'll look into it!

Reply:
Done!

Reply:
Well, I added a suggestion to add the lint on the match, I didn't find how to add a suggestion to remove the lint on the arm

Reply:
You need to suggest an empty string in place of the current lint attribute's span.

Reply:
Hmm, I can get the span for the lint level, but it's not the span of the whole attribute; removing it does:
```diff
-         #[deny(non_exhaustive_omitted_patterns)]
+         #[deny()]
           _ => {}
```
It seems `lint_level_at_node` does not return anything else than this span to identify where the lint was set. Can I use the span to get to the span of the whole attribute somehow? Or use the arm `HirId`? It's not helping that rustc internal docs are missing atm :'(

Reply:
Turns out that even with a better span, it's pretty hard to make clear that I'm suggesting to remove the line. Instead I added a "remove this" span label, it's a lot clearer ^^